proto/public/uber/cadence/api/v1/service_worker.proto

```diff
@@ -0,0 +1,323 @@
+  // to heartbeat within 'heartbeatTimeoutSeconds' interval for the ActivityTask, then it will be marked as timedout and
+  // 'ActivityTaskTimedOut' event will be written to the workflow history.  Calling 'RecordActivityTaskHeartbeat' will
+  // fail with 'EntityNotExistsError' in such situations.  Use the 'taskToken' provided as response of
+  // PollForActivityTask API call for heartbeating.
+  rpc RecordActivityTaskHeartbeat(RecordActivityTaskHeartbeatRequest) returns (RecordActivityTaskHeartbeatResponse);
+
+  // RecordActivityTaskHeartbeatByID is called by application worker while it is processing an ActivityTask.  If worker
+  // fails to heartbeat within 'heartbeatTimeoutSeconds' interval for the ActivityTask, then it will be marked as
+  // timed out and 'ActivityTaskTimedOut' event will be written to the workflow history.
+  // Calling 'RecordActivityTaskHeartbeatByID' will fail with 'EntityNotExistsError' in such situations.  Instead of
+  // using 'taskToken' like in RecordActivityTaskHeartbeat, use Domain, WorkflowID and ActivityID.
+  rpc RecordActivityTaskHeartbeatByID(RecordActivityTaskHeartbeatByIDRequest) returns (RecordActivityTaskHeartbeatByIDResponse);
+
+
+  // RespondQueryTaskCompleted is called by application worker to complete a QueryTask (which is a DecisionTask for query)
+  // as a result of 'PollForDecisionTask' API call. Completing a QueryTask will unblock the client call to 'QueryWorkflow'
+  // API and return the query result to client as a response to 'QueryWorkflow' API call.
+  rpc RespondQueryTaskCompleted(RespondQueryTaskCompletedRequest) returns (RespondQueryTaskCompletedResponse);
+
+  // RequestCancelWorkflowExecution is called by application worker when it wants to request cancellation of a workflow
+  // instance. It will result in a new 'WorkflowExecutionCancelRequested' event being written to the workflow history
+  // and a new DecisionTask created for the workflow instance so new decisions could be made. It fails with
+  // 'EntityNotExistsError' if the workflow is not valid anymore due to completion or doesn't exist.
+  rpc RequestCancelWorkflowExecution(RequestCancelWorkflowExecutionRequest) returns (RequestCancelWorkflowExecutionResponse);
+}
+
+message PollForDecisionTaskRequest {
+  string domain = 1;
+  TaskList task_list = 2;
+  string identity = 3;
+  string binary_checksum = 4;
+}
+
+message PollForDecisionTaskResponse {
+  bytes task_token = 1;
+  WorkflowExecution workflow_execution = 2;
+  WorkflowType workflow_type = 3;
+  int64 previous_started_event_id = 4;
+  int64 started_event_id = 5;
+  int64 attempt = 6;
+  int64 backlog_count_hint = 7;
+  History history = 8;
+  bytes next_page_token = 9;
+  WorkflowQuery query = 10;
+  TaskList workflow_execution_task_list = 11;
+  google.protobuf.Timestamp scheduled_time = 12 [(gogoproto.stdtime) = true];
+  google.protobuf.Timestamp started_time = 13 [(gogoproto.stdtime) = true];
+  map<string, WorkflowQuery> queries = 14;
```

Comment:
both "query" and "queries" is fairly odd, especially since the RespondDecisionTaskCompletedRequest has only a query_results map.  is "query" perhaps just un-cleaned-up thrift legacy?

Reply:
This is not yet legacy, but it might be a good timing to drop this.

Reply:
This is for consistent query. @andrewjdawson2016 has more context.